README
== Synopsis
A C extension to ruby enabling fast escape/unescape of
URL Encoded Strings

== Usage

 URLEscape.unescape("a%20string")
 URLEscape.escape("a string")


CHANGELOG
[0611913 | Wed Jun 24 19:37:33 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * updated changelog/authors for release

[41e2892 | Wed Jun 24 18:16:44 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * added more RELEASE doc info

[8107148 | Wed Jun 24 18:09:12 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * Version 2009.06.24 - Initial Public Release

[471517d | Wed Jun 24 18:07:53 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * added copyright to the .rb files, and a LICENSE

[fcfd320 | Wed Jun 24 18:03:07 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * fixed copyright task

[cdd6fa3 | Wed Jun 24 18:02:04 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * added new copyright

[e2906c0 | Wed Jun 24 17:36:10 UTC 2009] Jayson Vaughn <jayson@onedrop.(none)>

  * Added myself as an author.  kinda an ego commit

[2c45d0a | Wed Jun 24 17:31:28 UTC 2009] Kevin Berry <KevinBerry@nrs.us>

  * modified the RBConfig search to take into account suffixed ruby binaries

[411df7a | Wed Jun 24 17:13:50 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * removed gemspec from git, added partial release file

[977c411 | Wed Jun 24 17:01:20 UTC 2009] Jayson Vaughn <jayson@onedrop.(none)>

  * Corrected typo to set GEMSPEC.version properly

[4f5fff3 | Wed Jun 24 16:56:58 UTC 2009] Jayson Vaughn <jayson@onedrop.(none)>

  * Updated Rakefile to use the jruby URLEscape library if using jruby

[4b294eb | Wed Jun 24 16:02:28 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * removed strings which throw errors on 1.9

[b238261 | Wed Jun 24 15:53:22 UTC 2009] Jayson Vaughn <jayson@onedrop.(none)>

  * changed hex_table to be upper case, changed specs to test against hex represented with upper case chars

[3fe98dd | Wed Jun 24 15:43:05 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * um, rack does not work as expected, added tests to check rack::utils against our strings, unexpected behavior in 1.9.1

[994c06a | Wed Jun 24 15:35:29 UTC 2009] Jayson Vaughn <jayson@onedrop.(none)>

  * added java version and spec changes to support it

[0a51701 | Wed Jun 24 02:05:48 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * fixed big_bench

[39d6b2c | Tue Jun 23 23:24:32 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * added big_bench.rb for a real-life test of the condition.  on 1.9 (linux X86) the C does not make as much a difference. TODO: test on other platforms

[97949bb | Tue Jun 23 22:54:41 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * Updated build task for ruby 1.9.1 on windows

[19ea12f | Tue Jun 23 20:10:47 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * fixed assertions on benchmarks

[b08d38f | Tue Jun 23 19:58:26 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * Updated build task to check for common issues when building on windows

[9e594c4 | Tue Jun 23 19:04:51 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * Fixed the bacon.rake job so it works correctly in unix.

[3899455 | Tue Jun 23 18:50:02 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * * Modified build task to allow for windows

    * Modified bacon to allow for windows

[fc5a533 | Wed Jun 24 06:56:46 UTC 2009] Michael Fellinger <m.fellinger@gmail.com>

  * Remove some debugging output

[43332a6 | Wed Jun 24 06:53:44 UTC 2009] Michael Fellinger <m.fellinger@gmail.com>

  * Make rake build

[37c8c6c | Tue Jun 23 16:42:00 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * Added plus and space to be static as well.

[06481df | Tue Jun 23 16:40:10 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * Fixed issue with rubygems stepping on the extension in ruby >= 1.8.7

[169462c | Tue Jun 23 16:24:11 UTC 2009] TJ Vanderpoel <bougy.man@gmail.com>

  * changed load order to fix weirdness in 1.8.7

[472af85 | Wed Jun 24 06:18:18 UTC 2009] Michael Fellinger <m.fellinger@gmail.com>

  * Use block chdir

[7acd33d | Tue Jun 23 16:12:45 UTC 2009] TJ Vanderpoel <bougy.man@gmail.com>

  * changed require for weird breakage on 1.8.7

[cbb73f7 | Tue Jun 23 14:03:46 UTC 2009] TJ Vanderpoel <TJ Vanderpoel tj@rubyists.com>

  * cleaned up bench spec

[a2db063 | Tue Jun 23 00:53:22 UTC 2009] bougyman <bougyman@gotdebt.nationwide-recovery.com>

  * added rack dependency

[ca11d2b | Mon Jun 22 22:55:27 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * Started on Rakefile support for Windows

[c8dc880 | Mon Jun 22 21:44:05 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * Reordered variable declaration so that MSVC will compile the code

[0e712a1 | Mon Jun 22 18:38:45 UTC 2009] TJ Vanderpoel <Jayson Vaughn jv@rubyists.com>

  * added authors, cleaned up list for TD

[1dc7ca8 | Mon Jun 22 18:36:47 UTC 2009] TJ Vanderpoel <Jayson Vaughn jv@rubyists.com>

  * added bench.rb to spec files

[b0b4c74 | Mon Jun 22 18:35:26 UTC 2009] TJ Vanderpoel <Jayson Vaughn jv@rubyists.com>

  * added benchmark spec to make sure the speed increase is realized, and a clean rake task

[675acda | Mon Jun 22 18:21:36 UTC 2009] TJ Vanderpoel <Jayson Vaughn jv@rubyists.com>

  * Cleaned up variable names



[fa88077 | Mon Jun 22 15:48:14 UTC 2009] Evan Phoenix <ephoenix@engineyard.com>

  * Clean up new escape code

[a0509c6 | Mon Jun 22 17:08:32 UTC 2009] TJ Vanderpoel <Jayson Vaughn jv@rubyists.com>

  * added conditional for :force_encoding (for 1.8.6) in spec

[11dda51 | Mon Jun 22 16:58:35 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>

  * Fixed specs to work with ruby 1.8

[cc043f9 | Mon Jun 22 16:46:50 UTC 2009] TJ Vanderpoel <Jayson Vaughn jv@rubyists.com>

  * allocate 3 bytes earlier for speed increase

[fbdde47 | Mon Jun 22 16:38:36 UTC 2009] TJ Vanderpoel <Jayson Vaughn jv@rubyists.com>

  * removed gemspec from MANIFEST

[0a4b914 | Mon Jun 22 16:37:14 UTC 2009] TJ Vanderpoel <Jayson Vaughn jv@rubyists.com>

  * removed gemspec from manifest

[8051954 | Mon Jun 22 16:26:26 UTC 2009] Trey Dempsey <trey.dempsey@gmail.com>


ext/escape.c
/**
Copyright © 2009 Evan Phoenix and The Rubyists, LLC
Distributed under the terms of the MIT license.
See the LICENSE file which accompanies this software for the full text
*/

#include "ruby.h"

#ifndef RSTRING_PTR
#define RSTRING_PTR(obj) RSTRING(obj)->ptr
#endif

#ifndef RSTRING_LEN
#define RSTRING_LEN(obj) RSTRING(obj)->len
#endif

static const char space[] = { ' ' };
static const char plus[] =  { '+' };
static const char hex_table[][4] = {
  "%00", "%01", "%02", "%03", "%04", "%05", "%06", "%07",
  "%08", "%09", "%0A", "%0B", "%0C", "%0D", "%0E", "%0F",
  "%10", "%11", "%12", "%13", "%14", "%15", "%16", "%17",
  "%18", "%19", "%1A", "%1B", "%1C", "%1D", "%1E", "%1F",
  "%20", "%21", "%22", "%23", "%24", "%25", "%26", "%27",
  "%28", "%29", "%2A", "%2B", "%2C", "%2D", "%2E", "%2F",
  "%30", "%31", "%32", "%33", "%34", "%35", "%36", "%37",
  "%38", "%39", "%3A", "%3B", "%3C", "%3D", "%3E", "%3F",
  "%40", "%41", "%42", "%43", "%44", "%45", "%46", "%47",
  "%48", "%49", "%4A", "%4B", "%4C", "%4D", "%4E", "%4F",
  "%50", "%51", "%52", "%53", "%54", "%55", "%56", "%57",
  "%58", "%59", "%5A", "%5B", "%5C", "%5D", "%5E", "%5F",
  "%60", "%61", "%62", "%63", "%64", "%65", "%66", "%67",
  "%68", "%69", "%6A", "%6B", "%6C", "%6D", "%6E", "%6F",
  "%70", "%71", "%72", "%73", "%74", "%75", "%76", "%77",
  "%78", "%79", "%7A", "%7B", "%7C", "%7D", "%7E", "%7F",
  "%80", "%81", "%82", "%83", "%84", "%85", "%86", "%87",
  "%88", "%89", "%8A", "%8B", "%8C", "%8D", "%8E", "%8F",
  "%90", "%91", "%92", "%93", "%94", "%95", "%96", "%97",
  "%98", "%99", "%9A", "%9B", "%9C", "%9D", "%9E", "%9F",
  "%A0", "%A1", "%A2", "%A3", "%A4", "%A5", "%A6", "%A7",
  "%A8", "%A9", "%AA", "%AB", "%AC", "%AD", "%AE", "%AF",
  "%B0", "%B1", "%B2", "%B3", "%B4", "%B5", "%B6", "%B7",
  "%B8", "%B9", "%BA", "%BB", "%BC", "%BD", "%BE", "%BF",
  "%C0", "%C1", "%C2", "%C3", "%C4", "%C5", "%C6", "%C7",
  "%C8", "%C9", "%CA", "%CB", "%CC", "%CD", "%CE", "%CF",
  "%D0", "%D1", "%D2", "%D3", "%D4", "%D5", "%D6", "%D7",
  "%D8", "%D9", "%DA", "%DB", "%DC", "%DD", "%DE", "%DF",
  "%E0", "%E1", "%E2", "%E3", "%E4", "%E5", "%E6", "%E7",
  "%E8", "%E9", "%EA", "%EB", "%EC", "%ED", "%EE", "%EF",
  "%F0", "%F1", "%F2", "%F3", "%F4", "%F5", "%F6", "%F7",
  "%F8", "%F9", "%FA", "%FB", "%FC", "%FD", "%FE", "%FF"
};

static inline char hex(char digit) {
  if(digit >= '0' && digit <= '9') {
    return digit - '0';
  } else if(digit >= 'a' && digit <= 'f') {
    return 10 + digit - 'a';
  } else if(digit >= 'A' && digit <= 'F') {
    return 10 + digit - 'A';
  }

  return -1;
}

static VALUE unescape(VALUE self, VALUE str) {
  const char* buf;
  const char* bufend;
  VALUE outstr;

  StringValue(str);
  buf = RSTRING_PTR(str);
  bufend = buf + RSTRING_LEN(str);
  outstr = rb_str_buf_new(RSTRING_LEN(str));

  while(buf < bufend) {
    if(buf[0] == '%' && buf + 2 <= bufend) {
      char high = hex(buf[1]);
      char low =  hex(buf[2]);

      if(high >= 0 && low >= 0) {
        const char byte = low + (high << 4);
        rb_str_buf_cat(outstr, &byte, 1);
        buf += 3;
        continue;
      }
    }

    if(buf[0] == '+') {
      rb_str_buf_cat(outstr, space, 1);
    } else {
      rb_str_buf_cat(outstr, buf, 1);
    }
    buf++;
  }

  return outstr;
}


static inline int valid_literal(const char byte) {
  return ('A' <= byte && byte <= 'Z') ||
         ('a' <= byte && byte <= 'z') ||
         ('0' <= byte && byte <= '9') ||
         byte == '-' ||
         byte == '_' ||
         byte == '.';

}

static VALUE escape(VALUE self, VALUE str)
{
  char* buf;
  int len;
  VALUE outstr;
  int i;
  unsigned char byte_two, byte_three;

  StringValue(str);
  buf = RSTRING_PTR(str);
  len = RSTRING_LEN(str);
  outstr = rb_str_buf_new(len);

  for(i = 0; i < len;) {
    const unsigned char byte_one = buf[i];

    /* (UTF-8 escape, 0x0000-0x007F) */
    if(byte_one < 0x80) {
      if(valid_literal(byte_one)) {
        rb_str_buf_cat(outstr, buf+i, 1);
      } else if(byte_one == ' ') {
        // a + or %20 replacement (depending on const plus assignment)
        rb_str_buf_cat(outstr, plus, 1);
      } else { // It's ascii but needs encoding
        rb_str_buf_cat(outstr, hex_table[byte_one], 3);
      }
      i++;
      continue;
    }

    // Ok, there are UTF-8 prefix bytes, so we need to interpret
    // them.
    //
    // If we have at least one extra byte to consume
    if(i + 1 < len) {
      byte_two = buf[i + 1];

      /* (UTF-8 escape, 0x0080-0x07FF) */
      if(0xc0 <= byte_one && byte_one <= 0xdf && 0x80 <= byte_two && byte_two <= 0xbf) {
        rb_str_buf_cat(outstr, hex_table[byte_one], 3);
        rb_str_buf_cat(outstr, hex_table[byte_two], 3);
        i += 2;
        continue;

        // If we have at least two extra bytes to consume
      } else if(i + 2 < len) {
        byte_three = buf[i + 2];

        /* (UTF-8 escape, 0x0800-0xFFFF) */
        if(0xe0 == byte_one && 0x80 <= byte_two && byte_two <= 0xbf) {
          rb_str_buf_cat(outstr, hex_table[byte_one], 3);
          rb_str_buf_cat(outstr, hex_table[byte_two], 3);
          rb_str_buf_cat(outstr, hex_table[byte_three], 3);
          i += 3;
          continue;
        }
      }
    }

    /* (ISO Latin-1/2/? escape, 0x0080 - x00FF) */
    if(0x80 <= byte_one) {
      rb_str_buf_cat(outstr, hex_table[byte_one], 3);
    } else {
      // Well crap. Just throw it in I guess...
      rb_str_buf_cat(outstr, hex_table[byte_one], 3);
    }
    i++;
  }

  return outstr;
}

void Init_url_escape() {
  VALUE mod = rb_define_module("URLEscape");
  rb_define_singleton_method(mod, "unescape", unescape, 1);
  rb_define_singleton_method(mod, "escape",   escape,   1);
}
